intelmq/bin/intelmqctl.py

```diff
@@ -563,12 +563,38 @@ def botnet_stop(self, group=None):
         for bot_id in bots:
             self.bot_stop(bot_id, getstatus=False)
 
-        retval = 0
-        time.sleep(0.75)
-        for bot_id in bots:
-            botnet_status[bot_id] = self.bot_status(bot_id)[1]
-            if botnet_status[bot_id] not in ['stopped', 'disabled']:
-                retval = 1
+        # shallow copy of the list suffices
+        # only aliasing the list to ease reading the following
+        stopped_but_still_running_bots = bots
+
+        retries = getattr(self._parameters, 'stop_retry_limit', 5)
+
+        # parameters (default):
+        # - sleep 0.75 s with an increment of 0.1
+        # - at most 5 tries
+        # => sleep-ing at most 4.75 seconds
+        sleep_time = 0.75  # in seconds
+        for _ in range(retries):
+            # give the bots some time to terminate
+            time.sleep(sleep_time)
+            # update the botnet_status
+            for bot_id in stopped_but_still_running_bots:
+                botnet_status[bot_id] = self.bot_status(bot_id)[1]
+                if botnet_status[bot_id] in ['stopped', 'disabled']:
+                    stopped_but_still_running_bots.remove(bot_id)
```

Comment:
You iterate over a list while modifying it. That's now working in the way as you might expect:

```
>>> d = list(range(5))
>>> for x in d:
...     print(x)
...     if x == 2:
...         d.remove(x)
...         
0
1
2
4
```

That skips an element.

You need to iterate over a copy.